ee/clickhouse/queries/property_optimizer.py

```diff
@@ -0,0 +1,112 @@
+                return None, property_group
+            else:
+                return property_group, None
+
+        else:
+            # Top level type is AND.
+            # If all person properties, push them down
+            if self.using_only_person_properties(property_group):
+                return None, property_group
+            else:
+                # Mixed, look into each individual group
+                if isinstance(property_group.values[0], PropertyGroup):
+                    # multiple nested property groups
+                    # we care about selecting AND groups with person properties
+                    outer_property_group_values = []
+                    inner_property_group_values = []
+                    for group in property_group.values:
+                        assert isinstance(group, PropertyGroup)
+
+                        outer, inner = self._parse_property_groups(group)
+                        if outer:
+                            outer_property_group_values.append(outer)
+
+                        if inner:
+                            inner_property_group_values.append(inner)
+
+                    return (
+                        PropertyGroup(PropertyOperatorType.AND, outer_property_group_values),
+                        PropertyGroup(PropertyOperatorType.AND, inner_property_group_values),
+                    )
+
+                elif isinstance(property_group.values[0], Property):
+                    # list of properties that have been AND'ed, safe to push
+                    # all person properties down, rest go to outer group
+                    outer_property_values = []
+                    inner_property_values = []
+                    for property in property_group.values:
+                        assert isinstance(property, Property)
+
+                        if property.type == "person":
+                            inner_property_values.append(property)
+                        else:
+                            outer_property_values.append(property)
+                    return (
+                        PropertyGroup(PropertyOperatorType.AND, outer_property_values),
+                        PropertyGroup(PropertyOperatorType.AND, inner_property_values),
+                    )
+
+                else:
+                    raise ValidationError("Invalid property group values")
+
+    @staticmethod
+    def using_only_person_properties(property_group: PropertyGroup) -> bool:
+        if len(property_group.values) == 0:
+            return True
+
+        if isinstance(property_group.values[0], Property):
+            return all(property.type == "person" for property in property_group.values)
+
+        elif isinstance(property_group.values[0], PropertyGroup):
+            return all(
+                PropertyOptimizer.using_only_person_properties(group)
+                for group in cast(List[PropertyGroup], property_group.values)
+            )
+
+        else:
```

Comment:
If this is impossible, you can also just do if-else.

Reply:
Yep, most probably yes, just wanted it here first as a sanity check. Will trace all paths through which we can get here and then revisit

Reply:
I want to keep this, since this method can be called from anywhere, and it being defensive is better than it assuming it has correct inputs. Although, I don't expect this code path to ever be called (if we're doing things right xD 🤷 )